rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,648 @@
+Spanner or etcd. If fault tolerance isn't required, the user could configure
+state to be held in memory only.
+
+#### Leadership Transfer
+
+The master writes state to journal files so that the state can be recovered on
+restart. It is possible that a new master could be brought up while the old
+master is still running. If we aren't careful, this could result in corruption
+of the journal as both masters try to write to it.
+
+Ideally we could rely on a distributed coordination service such as ZooKeeper.
+However, this would add a significant burden to users who don't have access to a
+ZooKeeper cluster, and it would also require adding a new dependency on a
+ZooKeeper client.
+
+What TensorFlow does have is a FileSystem API. We will leverage this API to
+perform leadership transfer by creating empty files and inspecting file
+modification times.
+
+```
+files = list_directory(leadership_directory)
+if all_files_older_than(files, leadership_transfer_interval):
+  file = create_unique_file(leadership_directory);
+  if file_is_strictly_newest(file, leadership_directory):
+    become_leader()
+# Another master may be leader. Wait for some time before trying again.
+wait_random_interval()
+```
+
+The leader master will periodically write files to the leadership directory to
+indicate that it is still leading.
+
+The above scheme relies on the filesystem's create_file() and list() operations
+being strongly consistent . Users may opt to use a filesystem that doesn't
+support strong consistency, but they do so at the risk of two concurrently
+running masters thinking they are leader. Common filesystems such as POSIX,
+HDFS, and GCS support such strong consistency, but S3 does not.
+
+#### Caveats
+
+This section calls out caveats that users will need to be aware of when using
+the tf.data service.
+
+-   Due to the nature of dataset splitting, elements will not be processed in
+    the same order as they were in the pre-distributed dataset. If a dataset
+    relies on the order of the input files, the user's assumptions will be
+    violated when splitting causes each input worker to process only a subset of
+    the input files.
+-   If a particular dataset operation doesn't support splitting, it must be
+    moved after the part of the dataset which is distributed. Alternately, the
+    user could set num_tasks=1 to avoid the need for splitting, but this will
+    have a heavy performance cost since it only allows a single worker to
+    generate dataset elements. The most commonly used but unsupported datasets
+    are `from_generator` and `zip`.
+
+#### Framework Integration
+
+Many users interact with TensorFlow through a framework such as
+[TFX](https://www.tensorflow.org/tfx). A framework can make leveraging the
+tf.data service as simple as toggling a configuration boolean, triggering the
+framework to bring up tf.data service servers and add a
+`tf.data.experimental.service.distribute` transformation at the end of the
```

Comment:
Tooling could be built around the tf.data service so that it can automatically scale, but I wouldn't consider that part of the core tf.data service. To enable autoscaling, the tf.data service would report whether it could use more resources, and it is up to external tooling to start more tf.data servers.